consensus/core/src/core.rs

```diff
@@ -275,11 +284,34 @@ impl Core {
     /// Adds/processed all the newly `accepted_blocks`. We basically try to move the threshold clock and add them to the
     /// pending ancestors list.
     fn add_accepted_blocks(&mut self, accepted_blocks: Vec<VerifiedBlock>) {
-        // Advance the threshold clock. If advanced to a new round then send a signal that a new quorum has been received.
-        if let Some(new_round) = self
-            .threshold_clock
-            .add_blocks(accepted_blocks.iter().map(|b| b.reference()).collect())
+        // Get max round of accepted blocks. This will be equal to the threshold
+        // clock round, either by advancing the threshold clock round by being
+        // greater than current clock round or by equaling the current clock round.
+        let max_accepted_round = accepted_blocks
+            .iter()
+            .map(|b| b.round())
+            .max()
+            .unwrap_or(GENESIS_ROUND);
+        // Therefore the leader round for which proposals will wait will be max accepted round - 1
+        // or saturate to GENESIS_ROUND.
+        let accepted_proposal_leader_round = max_accepted_round.saturating_sub(1);
+
+        // Ignore checking for leader blocks with rounds less than the current
+        // threshold clock round - 1.
+        let proposal_leaders_exist = if accepted_proposal_leader_round
+            >= self.threshold_clock.get_round().saturating_sub(1)
         {
+            self.leaders_exist(accepted_proposal_leader_round)
+        } else {
+            false
+        };
```

Comment:
Yeah its just for metrics. Wanted to be able to calculate the time we are actually spent waiting for a leader, this will handle the case where the leader is obtained during the normal quorum aggregation time. Then we can calculate any time that elapses after that specifically waiting for a leader. I moved this part of the code out into this PR and added the reasonings behind it. So feel free to let me know if you think this change makes sense to move forward with.

https://github.com/MystenLabs/sui/pull/19655